File: components/service-standard/src/com/github/joerodriguez/sbng2ex/service/ServiceResponse.java
package com.github.joerodriguez.sbng2ex.service;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class ServiceResponse<T> {
    private List<ServiceError> errors;
    private T entity;

    public ServiceResponse() {
        this.errors = new ArrayList<>();
        this.entity = null;
    }

    public static <T> ServiceResponse<T> create(Consumer<ServiceResponse<T>> consumer) {
        ServiceResponse<T> response = new ServiceResponse<>();

        consumer.accept(response);

        return response;
    }

    public boolean isSuccess() {
        return errors.isEmpty();
    }

    public T getEntity() {
        return entity;
    }

    public List<ServiceError> getErrors() {
        return errors;
    }

    public void addError(ServiceError error) {
        this.errors.add(error);
    }

    public void commit(Supplier<T> supplier) {
        if (this.errors.isEmpty()) {
            this.entity = supplier.get();
        }
    }

    public static <T> ServiceResponse success(T entity) {
        return ServiceResponse.create((r) -> {
            r.commit(() -> entity);
        });
    }

    public void setEntity(T entity) {
        this.entity = entity;
    }

}

File: components/service-standard/src/com/github/joerodriguez/sbng2ex/service/ServiceResponse.kt
package com.github.joerodriguez.sbng2ex.service

import java.util.ArrayList

class ServiceResponse<T> {
    private val errors: MutableList<ServiceError>
    var entity: T? = null

    init {
        this.errors = ArrayList<ServiceError>()
        this.entity = null
    }

    val isSuccess: Boolean
        get() = errors.isEmpty()

    fun getErrors(): List<ServiceError> {
        return errors
    }

    fun addError(error: ServiceError) {
        this.errors.add(error)
    }

    fun apply(supplier: () -> T) {
        if (this.errors.isEmpty()) {
            this.entity = supplier.invoke()
        }
    }

    companion object {

        fun <T> create(consumer: (ServiceResponse<T>) -> Unit): ServiceResponse<T> {
            val response = ServiceResponse<T>()

            consumer.invoke(response)

            return response
        }

        fun <T> success(entity: T): ServiceResponse<T> {
            return ServiceResponse.create {
                it.apply({ entity })
            }
        }
    }

}
